Tvl pool staking

PieWol opened the pull request:
What does this PR do?
- Introduced the TotalValueLocked storage for nomination-pools. 
- introduced a slashing api in mock.rs 
- additional test for tracking a slashing event towards a pool without sub-pools
- migration for the nomination-pools (V6 to V7) with `VersionedMigration`

Why are these changes needed?
this is the continuation of the work by @kianenigma in this [PR](https://github.com/paritytech/substrate/pull/13319)

How were these changes implemented and what do they affect?
- It's an extra StorageValue that's modified whenever funds flow in or out of staking for any of the `bonded_account` of `BondedPools`
- The `PoolSlashed`event is now emitted even when no `SubPools` are found

Closes https://github.com/paritytech/polkadot-sdk/issues/155
KSM: HHEEgVzcqL3kCXgsxSfJMbsTy8dxoTctuXtpY94n4s8F4pS

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3504274

liamaharon commented on substrate/frame/nomination-pools/src/tests.rs:
```diff
@@ -3290,8 +3301,9 @@ mod withdraw_unbonded {
-					let mut x = UnbondingBalanceMap::get();
-					*x.get_mut(&default_bonded_account()).unwrap() /= 5;
+					let x = UnbondingBalanceMap::get();
+					// TODO:
```
👀

PieWol replied:
I removed this test, as there was no reason to introduce it. It didn't exist before this PR.

PieWol commented on substrate/frame/nomination-pools/src/migration.rs:
```diff
@@ -724,4 +724,103 @@ pub mod v5 {
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(_data: Vec<u8>) -> Result<(), TryRuntimeError> {
```
I totally agree, but I do wonder how you wold check for that without repeating the exact same calculation. For further checking it's also possible to check for keeping the upper bound of the sum of all `total_balance` of all `PoolMembers` but thats also not an exact check as you seem to wish for. Iterating through all the PoolMembers is sadly very heavy. Can you give me an idea on how to properly do the `post_upgrade` checks?

liamaharon replied:
`post_upgrade` / `pre_upgrade` are not executed on-chain, so it doesn't matter if they're heavy.

you can abstract the tvl calculation into a function if you'd like to reuse it between `OnRuntimeUpgrade` methods

PieWol replied:
thanks, done. Now both checks I mentioned are included in `post_upgrade`.

Ank4n replied:
This is probably not needed since try state checks already include them and we can run migration checks with the try-state checks which will verify again what you are doing in post_upgrade. I think just checking the storage item is set with a value is good enough and what we generally do in other migrations.

PieWol commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -2796,9 +2850,12 @@ impl<T: Config> Pallet<T> {
+			u256(current_balance)
+				.saturating_mul(u256(points))
+				// We check for zero above
+				.div(u256(current_points)),
```
That's a panic, but as the code is right now that can never happen. I have not modified the workings of this code. Only the type conversion of `current_points` was changed from `u128` to `u256`.

kianenigma replied:
what prompted you to convert `current_points` to u256? 

If it was done in `Balance` before, I presume it was becuase someone analyzed it and concluded that it is fine. 

It indeed seems like the outer division is safe and didn't need to be done in u256.

PieWol replied:
Indeed, it's an unnecessary conversion. I introduced it to fix the code after accidentally placing parenthesis at the wrong place. I will revert the change.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3519307

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3519309

PieWol commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -2087,10 +2140,12 @@ pub mod pallet {
 			// is destroying then `withdraw_unbonded` can be used.
 			ensure!(pool.state != PoolState::Destroying, Error::<T>::NotDestroying);
-			T::Staking::withdraw_unbonded(pool.bonded_account(), num_slashing_spans)?;
+			pool.withdraw_from_staking(num_slashing_spans)?;
```
thanks Liam 😄

PieWol commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3269,25 +3353,30 @@ impl<T: Config> sp_staking::OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pall
+					SubPoolsStorage::<T>::insert(pool_id, sub_pools);
+				},
+				None => {},
 			};
```
reworked that function and added the requested comments.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3536351

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3536349

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3536466

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3536467

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3536465

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3537643

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3538394

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3538504

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3538506

PieWol commented:
The storage migration works with `try-runtime` on a recent Kusama snapshot:

```[2023-09-03T17:39:55Z INFO  frame_support::migrations] Running NominationPools VersionedOnRuntimeUpgrade: version 5 to 6.
[2023-09-03T17:39:55Z INFO  runtime::nomination-pools] [19343161] 🏊‍♂️ Upgraded 149 pools with a TVL of 61550133549019336
[2023-09-03T17:39:56Z INFO  try-runtime::cli] TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = (5168150000000 ps, 18446744073709551615 byte), total weight = (2000000000000 ps, 18446744073709551615 byte) (258.41 %, 100.00 %).```

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3542546

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3543055

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3543056

kianenigma commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -1231,6 +1263,27 @@ impl<T: Config> BondedPool<T> {
+	///
+	/// Returns the result of [`T::Staking::withdraw_unbonded`]
+	fn withdraw_from_staking(&self, num_slashing_spans: u32) -> Result<bool, DispatchError> {
+		let bonded_account = self.bonded_account();
```
won't ask you to change, but here you are choosing readability over performance: `self.bonded_account` was already known at the call site of this function, and it is non-trivial to re-compute. Wanted to make sure you are aware of that.

PieWol replied:
The way I see it it's only known in one of the two usages.

kianenigma replied:
yeah, you are right.

kianenigma commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3263,19 +3352,29 @@ impl<T: Config> Pallet<T> {
+			tvl.saturating_reduce(total_slashed);
+		});
+
+		if let Some(mut sub_pools) = SubPoolsStorage::<T>::get(pool_id) {
```
```suggestion
		if let Some(mut sub_pools) = SubPoolsStorage::<T>::get(pool_id).defensive() {
```

PieWol replied:
Would you mind to tell me why this change should be introduced? I don't see why I would want to log an error when no `SubPool`'s are found.

PieWol replied:
I could see this making sense if the param  `slashed_unlocking` would be changed to `Option(&BTreeMap)`

PieWol replied:
I will put this in the next if let statement that tries to get the  `SubPool::with_era` as specified in the param `slashed_unlocking` because this one is actually always expected to be found.

Ank4n replied:
I think Kian meant that SubPools should always exist since we have already checked above that pool exists and there is some unlocking balance that has been slashed.

PieWol replied:
Consider this setup: 
A pool gets created and there will be an active bond. Now a slash happens towards the validator that this pool selected. 
There should only be active stake for this pool as no unbonding has happenend.
The slashing event/hook will contain the pool's accountID and enter the `on_slash` fn. 

I think now we would expect the regular `BondedPool` to be found and have an event emitted but there should not be any defensive error about a SubPool not being found as it's completely legal to have no SubPool without any prior unbonding.

PieWol replied:
> SubPools should always exist since we have already checked above that pool exists and there is some unlocking balance that has been slashed.

I think it's not guaranteed at this point that any unlocking balance has been slashed.

Ank4n replied:
Yes you are right, its not guaranteed if there has been no unbonding yet in the pool. Though if subpools don't exist, it would be good to assert `slashed_unlocking` is empty as well.

PieWol replied:
Okay, I would suggest I change `slashed_unlocking` to be an `Option(BTreeMap)` and just pass `None` into the `on_slash`. This would seem very intuitive to me and cleaner to do these explicit checks. What do you think?

PieWol replied:
Yes, good idea.

kianenigma reviewed (commented):
Thanks for the effort so far! 

I have two broad questions: 

1. Your comments and the `try_state` code indicate that the

> TVL must be equal to or less than the total balance of all PoolMembers

But it is not clear when this inaccuracy can exist? the TVL, since it is tracking slashes as well, should always be accurate from what I can see.

2. You have slightly changed the definition of TVL in this PR. Let me elaborate a bit on how pools work: Pools take some balance from the user, transfer it to the pool account, and issue/burn equivalent points in return. You can imagine it as a bridge: You lock DOTs, and you get points in return. The only functions in which this bridging happens is the `brun` and `issue`. 

```
DOT <-burn/issue--> Points
```

The point of TVL in my mind was to track this process. It should track the total amount of DOTs that have been exchanged for points in the pool pallet. 

If you agree with this, while your approach is also likely _equivalent_, it is _different_ from what I have in mind. You don't need to query staking before and after unbond. You only need to track successful calls to `burn` and `issue`. This is more or less what the original PR was doing. 

cc @Ank4n @gpestana would be good to get your eyes on this as well.

PieWol commented:
For my understanding the perfect definition of `value locked` is exactly that what I would get back as a the amount when quering `T::Staking::total_stake()`. It's the actively bonded and currently unbonding funds. So the *value* that is currently *locked* 

Thus the `TVL` of the nom-pools should just be in exact sync with the sum of all `bonded_accounts` of all nom-pools. Initially I also thought that this would be achievable by just doing operations on `issue` and `dissolve` but due to the fact that only active stake is read directly from staking but the `UnbondingPool`'s are stored individuall y inside the pallet this doesn't always stay in check due to the one extrinsic called `pool_withdraw_unbonded`. This fn removes all unlocking chunks in `T::Staking` for the `bonded_account` permissionlessly. 

With the effect of the `T::Staking::total_stake()` yielding afterwards less value for the `bonded_account` than you would expect when querying the `total_balance` of all `PoolMembers` because the nom-pool included `UnbondingPools` are not updated.

This is the explanation why I did it different. I was just happy this way as I could have everything stay in check with my opinionated definition of `TVL` that `T::Staking` already provides that I could also use as a way of verifying the `TVL` inside of `do_try_state`. Furthermore this is an easy and also exact migration.

This should also answer the first question. As `total_balance()` it's possible for the sum of `total_balance` of all PoolMembers to be higher than the `T::Staking::total_stake()` for the `BondedPool`.

I'm completely fine to just do it as you wish but I quickly wanted to write up my thoughts behind the change of concept. @kianenigma

PieWol commented:
I do understand though, that there is also the way of looking at the `TVL` as the `TVL` in this very pallet. Then I would still suggest to change the initial way of doing it as it's not needed to manipulate ´TVL` on `issue` and `dissolve` as it would lead to unnecessary operations whenever funds are transferred between (sub-) pools that cause a zero sum operation on the `TVL`. Instead it would be enough to only increase the `TVL` on bonding new funds and decrease it upon withdrawal aswell as slashing. Let me know what you think.

kianenigma commented:
>  Initially I also thought that this would be achievable by just doing operations on `issue` and `dissolve` but due to the fact that only active stake is read directly from staking but the UnbondingPool's are stored individually inside the pallet this doesn't always stay in check due to the one extrinsic called pool_withdraw_unbonded

Okay, I see a good point here and I will rephrase: The TVL that I suggested only tracks `active_stake` that is in the pallet, because once you unbond, it is already out of the TVL. Your definition has this as well. 

If your approach was significantly more complicated, I would perhaps argue that we should look at the scope where this is used and see which one is more suitable, but your approach is also fine. The context where we need this is as the two scenarios below: 

1. We want to mark the amount of funds that is locked in Pools, both active and unbonding, as `T::Currency::deactivate`, which will in turn make them effectless in the OpenGov curves. 
2. Possibly, UIs can also benefit from this. 

With that in mind, I am convinced that your definition of total value locked is better.

kianenigma commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3169,8 +3232,34 @@ impl<T: Config> Pallet<T> {
+				.reduce(|acc, total_balance| acc + total_balance)
+				.unwrap_or_default();
+
+			ensure!(
```
Bit worried that this is actually too slow, but this is a concern independent of this pallet and similar to https://github.com/paritytech/polkadot-sdk/issues/234

kianenigma commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3263,19 +3352,29 @@ impl<T: Config> Pallet<T> {
-			};
-			for (era, slashed_balance) in slashed_unlocking.iter() {
+		let Some(pool_id) = ReversePoolIdLookup::<T>::get(pool_account).defensive() else { return };
+		// As the slashed account belongs to a [`BondedPool`] the [`TotalValueLocked`] decreases and
```
wfiw these lines will not really be rendered in any rust-doc, so the linking is not effective.

PieWol replied:
I see, I'll remove the brackets.

kianenigma commented on substrate/frame/nomination-pools/src/migration.rs:
```diff
@@ -725,3 +725,126 @@ pub mod v5 {
+	pub struct VersionUncheckedMigrateV5ToV6<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> VersionUncheckedMigrateV5ToV6<T> {
+		fn calculate_tvl_by_total_stake() -> BalanceOf<T> {
+			BondedPools::<T>::iter()
```
fyi the typical attack on this is for an attacker to wait until this migration is _just about to happen_, then inflate this map. 

In Polkadot and Kusama this cannot happen, but you can still write it as a warning about the assumptions of this migration.

PieWol replied:
I added a warning to the code. Thanks for pointing it out.

kianenigma commented on substrate/frame/nomination-pools/src/migration.rs:
```diff
@@ -725,3 +725,126 @@ pub mod v5 {
+
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+			ensure!(
```
interesting. Harmless to have these, but what caused you to think you might mess with the decoding of these? the migration is only writing to `TotalValueLocked` after all, the rest is just reads.

Ank4n replied:
Yeah I think the below checks can be removed. The last migration was I think migrating one of the storage items and had these checks just to ensure it does not mess up anything. This migration does not touch any existing storages.

Ank4n replied:
Optionally these checks could be added to `try_state` if its not already to ensure all values are decodedable.

kianenigma commented on substrate/frame/nomination-pools/src/tests.rs:
```diff
@@ -6727,3 +6601,74 @@ mod commission {
 	}
 }
+mod slash {
+
```
```suggestion
```

kianenigma reviewed (approved):
Main logic LGTM module open comments. Second approver should look at the tests and mocks as well, as I only skimmed them.

PieWol commented:
Thank you for your detailed review @kianenigma, 
I have submitted changes for most of your comments and will address your remaining comments after I have received a reply from you on my follow-up questions.

Ank4n commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3169,8 +3229,34 @@ impl<T: Config> Pallet<T> {
 				pool is being destroyed and the depositor is the last member",
 			);
+
+			let expected_tvl: BalanceOf<T> = BondedPools::<T>::iter()
```
I think you are iterating inside [another loop](https://github.com/paritytech/polkadot-sdk/pull/1322/files#diff-fd1d9327e2dde740889545cb51e8e116c70b311d4318fd031cb95735ffdbb936L3152), probably mistake?

Also try to use the existing loop of BondedPools to calculate the TVL and add an ensure check afterwards.

PieWol replied:
yes, indeed mistake. 👍 
I will change this as you suggest.

PieWol replied:
What do you think about having them in separate iterations? The existing iterations over PoolMembers and BondedPools use fallible functions which could lead to having only some of the errors displayed, if the existing checks fail. I feel like the performance difference would be acceptable.

Ank4n commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3169,8 +3229,34 @@ impl<T: Config> Pallet<T> {
+				"TVL deviates from the actual sum of funds of all Pools."
+			);
+
+			let total_balance_members: BalanceOf<T> = PoolMembers::<T>::iter()
```
Similarly for this. The total balance could be calculated when iterate over `PoolMembers` earlier in the function.

Ank4n commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3263,31 +3349,40 @@ impl<T: Config> Pallet<T> {
+		// As the slashed account belongs to a `BondedPool` the `TotalValueLocked` decreases and
+		// an event is emitted.
+		TotalValueLocked::<T>::mutate(|tvl| {
+			tvl.saturating_reduce(total_slashed);
```
Should be `checked_sub` and a defensive assertion since slashed should never be higher than bonded amount (and hence never higher than TVL).

PieWol replied:
So currently any slash higher than `TVL` would result in `TVL` becoming zero but without any defensive error being provided notifying about the unreasonably high slash. 

As you now wish for a `checked_sub` what would be the expected outcome? Should it still be 0? Do you want to retain the original `TVL` and declare this slash as invalid? Whats the goal of this change? Just a defensive error in case the `total_slashed` would be larger than `TVL`?

PieWol replied:
Assuming that it's only about the defensive error I switched to `defensive_saturating_reduce`

Ank4n commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3263,31 +3349,40 @@ impl<T: Config> Pallet<T> {
+		if let Some(mut sub_pools) = SubPoolsStorage::<T>::get(pool_id) {
+			// set the reduced balance for each of the `SubPools`
+			slashed_unlocking.iter().for_each(|(era, slashed_balance)| {
+				if let Some(pool) = sub_pools.with_era.get_mut(era).defensive() {
```
Some of the slashed balance might be from an era that's been merged into no_era right? I am wondering if the older code was correct as well. If era is not found in a subpool with eras, the remaining non-slashed balance should be at the end slashed from subpool with no_era (I think). cc: @kianenigma @gpestana

PieWol replied:
I think a merge into a no_era pool is possible to minimize the storage taken by pool users that unbond but just never withdraw. If they would keep their own pool for every withdraw the storage gets bloated pretty quickly. So these unwanted pools get screwed in terms of having their own "point to balance ratio" and just get merged into this one big `SubPool::no_era` but only if they are way past the actual unbonding duration.

Furthermore in the current doc [this](https://github.com/paritytech/polkadot-sdk/blob/28aa4e94ea38a0647a97e4a77a61b08f02cf433e/substrate/frame/nomination-pools/src/lib.rs#L1537C1-L1539C95) explains why this should not be allowed to be not found. I guess the old code lacked the defensive error.

PieWol replied:
tldr: 
> Some of the slashed balance might be from an era that's been merged into no_era right? 
I dont think so.

PieWol replied:
@Ank4n I revisited this and stumbled across this [doc](https://github.com/paritytech/polkadot-sdk/blob/f1994c8690baa7f718f3fc5a351c8c6c9661e156/substrate/frame/staking/src/slashing.rs#L590C1-L593C28) comment in `slashing.rs`. 

The way I interpret this is that still `SubPools::no_era` will be untouched but any amount that remains of the slash might need to be deducted from the pool's `RewardPool`. This was neither accounted for in the previous code. 

Are these lines also where your concern originates from?

Ank4n commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3263,31 +3349,40 @@ impl<T: Config> Pallet<T> {
 		// anything here.
 		slashed_bonded: BalanceOf<T>,
 		slashed_unlocking: &BTreeMap<EraIndex, BalanceOf<T>>,
+		total_slashed: BalanceOf<T>,
```
The total slashed can be calculated from slashed_bonded and slashed_unlocking right? We could probably get rid of `slashed_bonded ` and emit `total_slashed` in the PoolSlashed event.

PieWol replied:
This would be possible but might also be a little confusing as there are already distinct events for a slash towards a `BondedPool` and an `UnbondingPool`. Having in mind that these events are emitted it would make a lot of sense to either stick to the way it is now or unite both events into one and have just `total_slashed` be included into the single event.

Ank4n reviewed (commented):
Thanks for the contribution. Generally looks really good ❤️.  Made some comments that should be fixed. I have not reviewed the tests yet but will do that soon.

Ank4n commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -1577,6 +1628,14 @@ pub mod pallet {
 
+	/// The sum of funds across all pools.
+	///
+	/// This might be higher but never lower than the actual sum of the currently unlocking and
```
I found the answer I was looking for [here](https://github.com/paritytech/polkadot-sdk/pull/1322#issuecomment-1708805742) but may be the next person will have the same question. Do you mind adding the explanation in the doc comment here?

PieWol replied:
yeah sorry for the Odyssey. I will add it to the doc.

PieWol commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -1231,6 +1263,27 @@ impl<T: Config> BondedPool<T> {
+		let prev_total = T::Staking::total_stake(&bonded_account.clone()).unwrap_or_default();
+		let outcome = T::Staking::withdraw_unbonded(bonded_account.clone(), num_slashing_spans);
+		let diff =
+			prev_total.saturating_sub(T::Staking::total_stake(&bonded_account).unwrap_or_default());
```
done, but again with `defensive_reducing_sub`.

PieWol commented:
@Ank4n thank you very much for your thorough review and uncovering some errors :)
I addressed all your comments by either commenting myself or just committing changes. Looking forward to your review on the tests.

Ank4n commented on substrate/frame/nomination-pools/src/mock.rs:
```diff
@@ -59,6 +60,19 @@ impl StakingMock {
+	///
+	/// Does not modify any [`SubPools::with_era`] of the pool as [`Default::default`] is passed for
+	/// `slashed_unlocking`.
+	pub fn slash_to(pool_id: PoolId, amount: Balance) {
```
I think it would be cleaner to combine `set_bonded_balance` and `slash_to` in one. They are doing pretty much same thing except the extra `Pools::on_slash`. You can check if the new balance in `set_bonded_balance` is reducing from before and call `Pools::on_slash` if so. Wdyt?

PieWol replied:
Forcing `on_slash` on every `set_bonded_balance` that reduces the `bonded_balance` decreases the ability to write more fine grained tests. For example in some tests `set_bonded_balance` is used to simply modify the `bonded_value` to make sure that this leads to certain behaviour. e.g. to test if the pool can be joined. I think basic tests like this should remain simple.

I think I should have just done a better job at combining new `TVL` checks with previous test.
Please let me know if this changes your mind.

Ank4n commented on substrate/frame/nomination-pools/src/tests.rs:
```diff
@@ -3239,179 +3248,29 @@ mod pool_withdraw_unbonded {
-	use sp_runtime::bounded_btree_map;
-
-	#[test]
-	fn withdraw_unbonded_works_against_slashed_no_era_sub_pool() {
```
Is this test not relevant anymore?

PieWol replied:
No, I introduced it but realized it doesn't make sense. There will never be slashes against the no_era pool.

PieWol replied:
Thats right. No idea how that happened. I will restore it.

PieWol commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -3169,13 +3228,28 @@ impl<T: Config> Pallet<T> {
+			"TVL deviates from the actual sum of funds of all Pools."
+		);
+
+		ensure!(
+			TotalValueLocked::<T>::get() <= total_balance_members,
+			"TVL must be equal to or less than the total balance of all PoolMembers."
+		);
```
Somebody could troll though and send funds to the pools account. Once thats done this try_state is cooked. I could introduce a test though to showcase this, if you would appreciate it.

PieWol replied:
I think even the ED is included in `total_balance` of the pool's creator. I have now created this test to showcase the difference. [dev@b4e9966](https://github.com/paritytech/polkadot-sdk/pull/1322/commits/b4e9966db257a19ccb8c641facb247e55d0a79b5)

PieWol commented on substrate/frame/nomination-pools/src/mock.rs:
```diff
@@ -115,11 +132,13 @@ impl sp_staking::StakingInterface for StakingMock {
-		x.remove(&who);
-		UnbondingBalanceMap::set(&x);
+		let mut unbonding_map = UnbondingBalanceMap::get();
+		let staker_map = unbonding_map.get_mut(&who).ok_or("not a staker")?;
```
good point.

PieWol commented on substrate/frame/nomination-pools/src/mock.rs:
```diff
@@ -104,8 +118,11 @@ impl sp_staking::StakingInterface for StakingMock {
+		let unlocking_at = era + Self::bonding_duration();
 		let mut y = UnbondingBalanceMap::get();
-		*y.entry(*who).or_insert(Self::Balance::zero()) += amount;
+		y.entry(*who).or_insert(Default::default()).push((unlocking_at, amount));
```
It's fine for now and I'll leave this for somebody to change who needs it changed.

Ank4n reviewed (approved):
Some small comments but otherwise looks good to me. 

Please also update the branch with latest upstream changes. Also few CI jobs are failing, try to fix them.

PieWol commented on substrate/frame/nomination-pools/src/tests.rs:
```diff
@@ -3264,12 +3264,47 @@ mod pool_withdraw_unbonded {
+
+			// The difference between TVL and member_balance is exactly the difference between
+			// `total_stake` and the `free_balance`.
+			// This relation is not guaranteed in the wild as arbitrary transfers towards
```
Any transfer that increases the free balance of the pool's bonded account that doesn't originate from the nom-pool pallet is not accounted for in the sum of `total_balance` of all pool members, nor in the TVL. 

If such a transfer would happen then free balance would be increased whereas sum of total_balance and TVL would remain the same. If we were now to show the relation as its done in this test it wouldn't work out because the difference between `total_stake` and `free_balance` is larger than the difference between `TVL` and sum of `total_balance` for all pool members. Hence the disclaimer that this test assumes things that may not hold up in the wild.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3855019

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3854990

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3855134

PieWol commented:
@Ank4n @kianenigma I have adressed the latest comments, merged master recently and improved the CI failures. Are we ready to merge? 👀

Ank4n commented on substrate/frame/nomination-pools/src/migration.rs:
```diff
@@ -37,6 +37,94 @@ pub mod versioned_migrations {
+///
+/// WARNING: This migration works under the assumption that the [`BondedPools`] cannot be inflated
+/// arbitrarily. Otherwise this migration could fail due to too high weight.
+pub mod v7 {
```
This can be made private and `VersionCheckedMigrateV6ToV7 ` moved to the public module `versioned_migrations`.

More context [here](https://github.com/paritytech/polkadot-sdk/blob/222ed3acba8b27d571a2a85d1678cf481f17d042/substrate/frame/examples/storage-migrations/src/migrations/v1.rs#L39-L157).

PieWol replied:
I see, thanks for the finishing touch 🙏

Ank4n commented:
bot fmt

kianenigma commented:
/tip medium

kianenigma commented:
/tip medium
